### PR TITLE
Refactor Leader key feature

### PR DESCRIPTION
**Pull request by fauxpark**
## Description

This is quite a large and fairly invasive rework of the Leader key. Instead of hooking into `matrix_scan_user()`, leader sequences are now defined in the `leader_end_user()` callback. The `LEADER_EXTERNS()` and `LEADER_DICTIONARY()` macros are now extinct, and the core functionality has been moved out of `process_leader.[ch]` into a dedicated `leader.[ch]`. The documentation has also been significantly improved.

I can't guarantee the changes to the user code still work, but most of the heavy lifting has already been done here, and I suspect some of these keymaps probably didn't compile in the first place.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by fauxpark** [comment]
`LEADER_NO_TIMEOUT` doesn't appear to work properly in develop:

`QK_LEAD` ... wait more than 300ms ... `a`, `s`
just prints `s` instead of doing the sequence.

I'll need to investigate some more to see if I can fix it.

**Comment by fauxpark** [comment]
It appears to work if you have `LEADER_PER_KEY_TIMING` also defined. But it should really be fine on its own.

File: quantum/leader.c [hunk 1]
```diff
@@ -0,0 +1,98 @@
+#include "leader.h"
```
Comment (drashna) [inline quantum/leader.c:1]:
```suggestion
// Copyright 2023 QMK
// SPDX-License-Identifier: GPL-2.0-or-later

#include "leader.h"
```

File: quantum/leader.h [hunk 2]
```diff
@@ -0,0 +1,116 @@
+#include <stdbool.h>
```
Comment (drashna) [inline quantum/leader.h:1]:
```suggestion
// Copyright 2023 QMK
// SPDX-License-Identifier: GPL-2.0-or-later

#include <stdbool.h>
```

File: tests/leader/config.h [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#pragma once
```
Comment (drashna) [inline tests/leader/config.h:1]:
```suggestion
// Copyright 2023 QMK
// SPDX-License-Identifier: GPL-2.0-or-later

#pragma once
```

File: tests/leader/leader_no_initial_timeout/config.h [hunk 4]
```diff
@@ -0,0 +1,5 @@
+#pragma once
```
Comment (drashna) [inline tests/leader/leader_no_initial_timeout/config.h:1]:
```suggestion
// Copyright 2023 QMK
// SPDX-License-Identifier: GPL-2.0-or-later

#pragma once
```

File: tests/leader/leader_no_initial_timeout/test_leader_no_initial_timeout.cpp [hunk 5]
```diff
@@ -0,0 +1,45 @@
+#include "keyboard_report_util.hpp"
```
Comment (drashna) [inline tests/leader/leader_no_initial_timeout/test_leader_no_initial_timeout.cpp:1]:
```suggestion
// Copyright 2023 QMK
// SPDX-License-Identifier: GPL-2.0-or-later

#include "keyboard_report_util.hpp"
```

File: tests/leader/leader_per_key_timeout/test_leader_per_key_timeout.cpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
+#include "keyboard_report_util.hpp"
```
Comment (drashna) [inline tests/leader/leader_per_key_timeout/test_leader_per_key_timeout.cpp:1]:
```suggestion
// Copyright 2023 QMK
// SPDX-License-Identifier: GPL-2.0-or-later

#include "keyboard_report_util.hpp"
```

File: tests/leader/leader_sequences.c [hunk 7]
```diff
@@ -0,0 +1,23 @@
+#include "quantum.h"
```
Comment (drashna) [inline tests/leader/leader_sequences.c:1]:
```suggestion
// Copyright 2023 QMK
// SPDX-License-Identifier: GPL-2.0-or-later

#include "quantum.h"
```

File: tests/leader/leader_strict_key_processing/config.h [hunk 8]
```diff
@@ -0,0 +1,5 @@
+#pragma once
```
Comment (drashna) [inline tests/leader/leader_strict_key_processing/config.h:1]:
```suggestion
// Copyright 2023 QMK
// SPDX-License-Identifier: GPL-2.0-or-later

#pragma once
```

File: tests/leader/leader_strict_key_processing/test_leader_strict_key_processing.cpp [hunk 9]
```diff
@@ -0,0 +1,42 @@
+#include "keyboard_report_util.hpp"
```
Comment (drashna) [inline tests/leader/leader_strict_key_processing/test_leader_strict_key_processing.cpp:1]:
```suggestion
// Copyright 2023 QMK
// SPDX-License-Identifier: GPL-2.0-or-later

#include "keyboard_report_util.hpp"
```

File: tests/leader/test_leader.cpp [hunk 10]
```diff
@@ -0,0 +1,221 @@
+#include "keyboard_report_util.hpp"
```
Comment (drashna) [inline tests/leader/test_leader.cpp:1]:
```suggestion
// Copyright 2023 QMK
// SPDX-License-Identifier: GPL-2.0-or-later

#include "keyboard_report_util.hpp"
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Review by tzarc — APPROVED** [review]

**Pull request merged**